testsuite/smoke-test/src/verifying_client.rs

```diff
@@ -0,0 +1,239 @@
+        let send_v = self.verifying_client.batch(batch);
+
+        tokio::join!(send_nv, send_v)
+    }
+}
+
+/// Running a request against a `VerifyingClient` and non-verifying `Client`
+/// at the same state version should always return the same response.
+fn assert_responses_equal(
+    recv_nv: Result<Response<MethodResponse>>,
+    recv_v: Result<Response<MethodResponse>>,
+) {
+    match (recv_nv, recv_v) {
+        (Ok(resp_nv), Ok(resp_v)) => {
+            let (inner_nv, state_nv) = resp_nv.into_parts();
+            let (inner_v, state_v) = resp_v.into_parts();
+
+            assert_eq!(state_nv.chain_id, state_v.chain_id);
+            assert_eq!(inner_nv, inner_v);
+        }
+        (Err(_), Err(_)) => (),
+        (recv_nv, recv_v) => {
+            panic!(
+                "client handling mismatch:\n\
+                 1. non-verifying client response: {:?}\n\
+                 2.     verifying client response: {:?}",
+                recv_nv, recv_v,
+            );
+        }
+    }
+}
+
+/// Running a batch of requests against a `VerifyingClient` and non-verifying
+/// `Client` at the same state version should always return the same responses.
+fn assert_batches_equal(
+    recv_nv: Result<Vec<Result<Response<MethodResponse>>>>,
+    recv_v: Result<Vec<Result<Response<MethodResponse>>>>,
+) {
+    let (batch_nv, batch_v) = match (recv_nv, recv_v) {
+        (Ok(batch_nv), Ok(batch_v)) => (batch_nv, batch_v),
+        (Err(_), Err(_)) => return,
+        (recv_nv, recv_v) => {
+            panic!(
+                "client batch handling mismatch:\n\
+                 1. non-verifying client batch response: {:?}\n\
+                 2.     verifying client batch response: {:?}",
+                recv_nv, recv_v,
+            );
+        }
+    };
+
+    if batch_nv.len() != batch_v.len() {
+        panic!(
+            "clients returned different batch sizes: {} vs {}\n\
+             1. non-verifying client batch: {:?}\n\
+             2.     verifying client batch: {:?}",
+            batch_nv.len(),
+            batch_v.len(),
+            batch_nv,
+            batch_v,
+        );
+    }
+
+    for (resp_nv, resp_v) in batch_nv.into_iter().zip(batch_v.into_iter()) {
+        assert_responses_equal(resp_nv, resp_v);
+    }
+}
+
+/// A `Strategy` for generating a `MethodRequest` at or before the given version
+/// and querying from the given set of accounts.
+fn arb_request(
+    accounts: &[AccountAddress],
+    current_state_version: u64,
+) -> impl Strategy<Value = MethodRequest> {
+    let arb_account = select(accounts.to_owned());
+    let arb_version = prop_oneof! [
+        5 => Just(current_state_version),
+        // query a historical version
+        20 => 0u64..current_state_version,
+        // occasionally choose an invalid version
+        1 => Just(u64::MAX / 2),
+    ];
+    (arb_account, arb_version)
+        .prop_map(|(address, version)| MethodRequest::get_account_by_version(address, version))
```

Comment:
So as we add more methods to the verifying client we can add more things here.

Reply:
Yeah that's the idea, generate different request types here as we add support for them